src/Acuminator/Acuminator.Analyzers/StaticAnalysis/PXGraphCreationForBqlQueries/PXGraphCreationForBqlQueriesAnalyzer.cs

```diff
@@ -0,0 +1,165 @@
+	public partial class PXGraphCreationForBqlQueriesAnalyzer : PXDiagnosticAnalyzer
+	{
+		public const string IdentifierNamePropertyPrefix = "IdentifierName";
+		public const string IsGraphExtensionPropertyPrefix = "IsGraphExtension";
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Descriptors.PX1072_PXGraphCreationForBqlQueries);
+
+		internal override void AnalyzeCompilation(CompilationStartAnalysisContext compilationStartContext, PXContext pxContext)
+		{
+			compilationStartContext.RegisterCodeBlockAction(c => AnalyzeCodeBlock(c, pxContext));
+		}
+
+		private void AnalyzeCodeBlock(CodeBlockAnalysisContext context, PXContext pxContext)
+		{
+			context.CancellationToken.ThrowIfCancellationRequested();
+
+			// Get body from a method or property
+			CSharpSyntaxNode body = GetBody(context.CodeBlock);
+			if (body == null) return;
+
+			// Collect all PXGraph-typed method parameters passed to BQL queries
+			var walker = new BqlGraphArgWalker(context.SemanticModel, pxContext);
+			body.Accept(walker);
+			if (walker.GraphArguments.IsEmpty) return;
+
+			// Collect all available PXGraph instances (@this, method parameters, local variables)
+			var existingGraphs = GetExistingGraphInstances(body, context.SemanticModel, pxContext);
+			if (existingGraphs.IsEmpty) return;
+
+			// Determine if available PXGraph instance is used outside of BQL queries
+			var usedGraphs = GetSymbolUsages(body, existingGraphs, context.SemanticModel, walker.GraphArguments)
+				.ToImmutableHashSet();
+			var availableGraphs = existingGraphs.Except(usedGraphs).ToArray();
+
+			// Analyze each PXGraph-typed parameter in BQL queries
+			foreach (var graphArgSyntax in walker.GraphArguments)
+			{
+				var instantiationType = graphArgSyntax.GetGraphInstantiationType(context.SemanticModel, pxContext);
+
+				// New PXGraph() / new TGraph() / PXGraph.CreateInstance<TGraph> are reported at all times
+				// All other usages are reported only if:
+				// 1. There is at least one existing PXGraph instance available
+				// 2. PXGraph parameter is not used in any way because its modifications might affect the BQL query results
+				if (instantiationType != GraphInstantiationType.None)
+				{
+					context.ReportDiagnostic(Diagnostic.Create(Descriptors.PX1072_PXGraphCreationForBqlQueries,
+						graphArgSyntax.GetLocation(),
+						CreateDiagnosticProperties(availableGraphs, pxContext)));
+				}
+				else if (availableGraphs.Length > 0 && context.SemanticModel.GetSymbolInfo(graphArgSyntax).Symbol is ILocalSymbol localVar 
+				                                    && !usedGraphs.Contains(localVar))
+				{
+					context.ReportDiagnostic(Diagnostic.Create(Descriptors.PX1072_PXGraphCreationForBqlQueries,
+						graphArgSyntax.GetLocation(),
+						CreateDiagnosticProperties(availableGraphs.Where(g => !Equals(g, localVar)), pxContext)));
+				}
+			}
+		}
+
+		private CSharpSyntaxNode GetBody(SyntaxNode codeBlock)
```

Comment:
@dropsonic this method could be shared. I suggest to move it to some syntax utils helper

Reply:
No objections.